File: src/main/java/com/bruno/WordMatcher.java
package com.bruno;

import com.bruno.matchers.Matcher;
import com.bruno.matchers.MatcherFactory;

import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;
import java.util.function.Predicate;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class WordMatcher {

    private Language language;
    private Matcher matcher;

    private int limit = 30;

    public WordMatcher(Language language, Matcher matcher) {
        this.language = language;
        this.matcher = matcher;
    }

    public WordMatcher setLimit(int limit) {
        this.limit = limit;

        return this;
    }

    public Stream<String> streamMatches(String word) throws URISyntaxException, IOException {
        Predicate<String> predicate = MatcherFactory.getInstance().getMatcher(this.matcher, word);

        URI fileURI = Application.class
                .getClassLoader()
                .getResource(this.language.getPath()).toURI();

        return Files
                .lines(Paths.get(fileURI))
                .parallel()
                .filter(predicate)
                .map(String::toLowerCase)
                .distinct()
                .limit(this.limit);
    }

    public List<String> getMatches(String word) throws URISyntaxException, IOException {
        return streamMatches(word).collect(Collectors.toList());
    }
}

File: src/main/java/com/bruno/WordMatcher.kt
package com.bruno

import com.bruno.matchers.Matcher
import com.bruno.matchers.getMatcher

import java.nio.file.Files
import java.nio.file.Paths
import java.util.stream.Collectors

class WordMatcher(private val language: Language, private val matcher: Matcher) {

    private var limit = 30L

    fun setLimit(limit: Long): WordMatcher {
        this.limit = limit

        return this
    }

    fun getMatches(word: String): List<String> {
        val predicate = getMatcher(this.matcher, word)

        val fileURI = Application::class.java
                .classLoader
                .getResource(this.language.path)!!.toURI()

        return Files
                .lines(Paths.get(fileURI))
                .parallel()
                .filter(predicate)
                .map { it.toLowerCase() }
                .distinct()
                .limit(this.limit)
                .collect(Collectors.toList())
    }
}
